src/include/sof/math/trig.h

```diff
@@ -137,17 +158,214 @@ static inline int16_t sin_fixed_16b(int32_t th_rad_fxp)
+ * with the angle in the range [-2*pi, 2*pi) and output has
+ * range in [-1.0 to 1.0]
+ * +------------------+-----------------+--------+------------+
+ * | thRadFxp	      |cdccexpth        |thRadFxp|   cdccexpth|
+ * +----+-----+-------+----+----+-------+--------+------------+
+ * |WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat    |
+ * +----+-----+-------+----+----+-------+--------+------------+
+ * | 32 | 28  |  1    | 32 | 15 |   1	| 4.28	 | 1.15       |
+ * +------------------+-----------------+--------+------------+
+ */
+static inline struct cmpx cmpx_exp_16b(int32_t th_rad_fxp)
+{
+	int32_t th_cdc_fxp;
+	int32_t sign;
+	int32_t b_yn;
+	int32_t xn;
+	struct cmpx cexp;
+
+	cordic_cfg type = EN_16B_CORDIC_CEXP;
+	/* compute coeff from angles*/
+	cordic_approx(th_rad_fxp, type, &sign, &b_yn, &xn, &th_cdc_fxp);
+	cexp = cmpx_cexp(sign, b_yn, xn, type);
+	/*return the complex result*/
+	return cexp;
+}
+
+/**
+ * CORDIC-based approximation of inverse cosine
+ * inverse cosine of cdc_acos_th based on a CORDIC approximation
+ * Inverse cosine angle values in rad
+ * Q2.30 cdc_acos_th , input value range [-1 to 1]
+ * Q2.30 th_rad_fxp, output value range [3.14159265346825 to 0]
+ * LUT size set type 31
+ * Error (max = 0.000000026077032), THD+N  = -157.948952635422842
+ */
+static inline int32_t acosine_fixed_32b(int32_t cdc_acos_th)
+{
+	int32_t saturatedUnaryMinus;
+	int32_t th_rad_fxp;
+	cordic_cfg type = EN_32B_CORDIC_ACOSINE;
+
+	saturatedUnaryMinus = cdc_acos_th;
+	if (saturatedUnaryMinus >= 0) {
+		th_rad_fxp = iscalarcordic(saturatedUnaryMinus, CORDIC_31B_TABLE_SIZE,
+					   cordic_ilookup, type);
+	} else {
+		if (saturatedUnaryMinus <= INT32_MIN)
+			saturatedUnaryMinus = INT32_MAX;
+		else
+			saturatedUnaryMinus = -saturatedUnaryMinus;
+
+	    th_rad_fxp =
+		PI_MUL2_Q4_28 - iscalarcordic(saturatedUnaryMinus, CORDIC_31B_TABLE_SIZE,
+		cordic_ilookup, type);
+	}
+	return th_rad_fxp;
+}
+
+/**
+ * CORDIC-based approximation of inverse cosine
+ * inverse cosine of cdc_acos_th based on a CORDIC approximation
+ * Inverse cosine angle values in rad
+ * Q2.30 cdc_acos_th , input value range [-1 to 1]
+ * Q2.30 th_rad_fxp, output value range [3.14159265346825 to 0]
+ * LUT size set type 15
+ * Error (max = 0.000000026077032), THD+N  = -157.948952635422842
+ */
+static inline int32_t acosine_fixed_16b(int32_t cdc_acos_th)
+{
+	int32_t saturatedUnaryMinus;
+	int32_t th_rad_fxp;
+	cordic_cfg type = EN_32B_CORDIC_ACOSINE;
+
+	saturatedUnaryMinus = cdc_acos_th;
+	if (saturatedUnaryMinus >= 0) {
+		th_rad_fxp = iscalarcordic(saturatedUnaryMinus, CORDIC_15B_ITABLE_SIZE,
+					   cordic_ilookup, type);
+	} else {
+		if (saturatedUnaryMinus <= INT32_MIN)
+			saturatedUnaryMinus = INT32_MAX;
+		else
+			saturatedUnaryMinus = -saturatedUnaryMinus;
+
+		th_rad_fxp =
+		PI_MUL2_Q4_28 - iscalarcordic(saturatedUnaryMinus, CORDIC_15B_ITABLE_SIZE,
+		cordic_ilookup, type);
+	}
+	return th_rad_fxp;
```

Comment:
This looks the same as acosine_fixed_32b ??

Reply:
function argument for number of iteration is different [31 and 15 ]!

Reply:
But Yes, this can be simplified to a single arc cosine routine. From double cordic iteration to  single iteration still maintaining the  Q2.30 format output.